[FIX] Nella fattura elettronica, compilare DatiPagamento anche senza termini di pagamento

SirPyTech opened the pull request:
Risolve https://github.com/OCA/l10n-italy/issues/3141, si può riaprire?

Incluso il refactoring https://github.com/OCA/l10n-italy/pull/4886/commits/c95e0849e83ad63a92df4df65821351e9add331f della PR https://github.com/OCA/l10n-italy/pull/4886 per poter aggregare entrambe le PR più facilmente.

SirPyTech committed:
[IMP] l10n_it_fatturapa_out: Set DatiPagamento without Payment Term

SirPyTech commented:
I test non partono nemmeno per https://github.com/OCA/l10n-italy/issues/4519#issuecomment-3083392333.

francesco-ooops commented:
> Risolve #3141, si può riaprire?

senti @monen17 😉

HekkiMelody reviewed (approved):
Code review, LGTM

HekkiMelody commented:
Anche functional review. Abbiamo cominciato a utilizzare questa modifica in produzione, con successo.

SirPyTech commented:
@DigitalAutomations @TheMule71 @francescapenso visto che avete partecipato nella issue, cosa ne pensate?

SirPyTech committed:
[REF] l10n_it_fatturapa_out: One page for e-invoice in invoice form

TheMule71 commented:
Col merge di #4893, puoi fare un rebase per far girare i test.

francesco-ooops commented:
/ocabot rebase

eLBati commented on l10n_it_fatturapa_out/migrations/16.0.1.7.0/pre-migrate.py:
```diff
@@ -0,0 +1,48 @@
+
+@openupgrade.migrate()
+def migrate(env, version):
+    precompute_new_computed_fields(env)
```
Questo per motivi di prestazioni?

SirPyTech replied:
Grazie della revisione :hugs:
Sì, ho seguito https://oca.github.io/OpenUpgrade/use_cases/field_fast_precomputation.html

eLBati reviewed (commented):
Per il resto LGTM

eLBati commented:
> Per il resto LGTM

A parte i test

SirPyTech committed:
[IMP] l10n_it_fatturapa_out_di: In tests, due date might be invisible

SirPyTech committed:
[IMP] l10n_it_fatturapa_out_rc: Adapt to new checks for DatiPagamento

If the due date is set, some data is mandatory to generate the DatiPagamento node in the e-invoice.
We don't need the DatiPagamento node in the self-invoice, so we can unset the due date.

HekkiMelody commented on l10n_it_fatturapa_out/readme/USAGE.md:
```diff
@@ -1,7 +1,7 @@
 
 > - Compilare la fattura con i dati necessari per l'esportazione: per
->   esempio, nella scheda "Allegati fattura elettronica"
+>   esempio, nelle schede "Allegati fattura elettronica" e "Fattura elettronica"
```
**question**: se ho interpretato bene, dopo questa modifica c'è solo una "scheda" `<page>` giusto?

SirPyTech replied:
Giusto, ho corretto e spostato la modifica alla documentazione nel commit di refactoring che effettivamente unisce le due pagine

HekkiMelody reviewed (approved):
Code review, LGTM.

HekkiMelody reviewed (approved):
Code review, LGTM

HekkiMelody commented:
@OCA/local-italy-developers utilizziamo con successo questa PR in prod ormai da svariati mesi. Grazie :pray:

eLBati commented:
/ocabot merge patch

HekkiMelody commented:
Grazie @eLBati !

eLBati commented on l10n_it_fatturapa_out/models/account.py:
```diff
@@ -55,35 +88,26 @@ def preventive_checks(self):
-                    _(
-                        "Invoice %(name)s fiscal payment term must be"
-                        " set for the selected payment term %(term)s"
+            if invoice.invoice_payment_term_id or invoice.invoice_date_due:
```
Questo controllo mi sembra troppo bloccante:

[Screencast from 25-11-2025 15:09:14.webm](https://github.com/user-attachments/assets/c19f789a-8fcc-4849-823e-0054d819e3e2)

HekkiMelody replied:
Ci può stare, ma il comportamento è consistente, era simile anche prima. Soltanto che prima il controllo c'era soltanto se si utilizzavano i termini di pagamento, e non la data direttamente in fattura.

Non credo che il flusso dovrebbe essere diverso tra le due modalità di inserimento della data, no?

eLBati replied:
L'idea era: se l'utente vuole utilizzare un termine di pagamento, allora per poterlo fare deve impostare i relativi campi per la fatturazione elettronica.
Ma se non lo vuole usare, non dovrebbe essere obbligato a impostarli sempre e comunque.

Inoltre, nella fatturazione ricorrente, qualora volesse utilizzare i termini di pagamento, le sarebbe sufficiente impostare i campi per l'XML nei termini di pagamento che vuole utilizzare.
Se invece non volesse impostare a priori i termini di pagamento, la fatturazione ricorrente sarebbe proprio bloccata.

Possiamo continuare su https://github.com/OCA/l10n-italy/pull/5009
